cpp/src/join/conditional_join.hpp

```diff
@@ -0,0 +1,78 @@
+/*
+ * Copyright (c) 2021, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#pragma once
+
+#include "join_common_utils.hpp"
+
+#include <cudf/ast/nodes.hpp>
+#include <cudf/table/table_view.hpp>
+
+#include <rmm/cuda_stream_view.hpp>
+#include <rmm/mr/device/per_device_resource.hpp>
+
+namespace cudf {
+namespace detail {
+
+constexpr int CONDITIONAL_JOIN_UNKNOWN_OUTPUT_SIZE = -1;
+
+/**
+ * @brief Computes the join operation between two tables and returns the
+ * output indices of left and right table as a combined table
+ *
+ * @param left  Table of left columns to join
+ * @param right Table of right  columns to join
+ * tables have been flipped, meaning the output indices should also be flipped
+ * @param JoinKind The type of join to be performed
+ * @param compare_nulls Controls whether null join-key values should match or not.
+ * @param stream CUDA stream used for device memory operations and kernel launches
+ *
+ * @return Join output indices vector pair
+ */
+std::pair<std::unique_ptr<rmm::device_uvector<size_type>>,
+          std::unique_ptr<rmm::device_uvector<size_type>>>
+conditional_join(table_view const& left,
+                 table_view const& right,
+                 ast::expression binary_predicate,
+                 null_equality compare_nulls,
+                 join_kind JoinKind,
+                 size_type output_size               = CONDITIONAL_JOIN_UNKNOWN_OUTPUT_SIZE,
```

Comment:
Use a `std::optional` here like we do with the hash join.

Reply:
Done.